classes/sample_set.php

```diff
@@ -17,31 +17,85 @@
 namespace tool_excimer;
 
 /**
- * Stores samples copied over from the profiler.
+ * Stores samples copied over from the profiler, to be used in a profile.
  *
  * @package    tool_excimer
  * @author     Jason den Dulk <jasondendulk@catalyst-au.net>
  * @copyright  2022 Catalyst IT
  * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
  */
-
 class sample_set {
     public $name;
-    public $samples = [];
     public $starttime;
 
+    public $samples = [];
+
+    public $samplelimit;
+
+    /** @var int If $filterrate is R, then only each Rth sample is recorded. */
+    private $filterrate = 1;
 
-    public function __construct($name, $starttime) {
+    /** @var int Internal counter to help with filtering. */
+    private $counter = 0;
+
+    /**
+     * Constructs the sample set.
+     *
+     * @param string $name
+     * @param float $starttime
+     * @param int $samplelimit
+     */
+    public function __construct(string $name, float $starttime, int $samplelimit) {
         $this->name = $name;
         $this->starttime = $starttime;
+        $this->samplelimit = $samplelimit;
     }
 
     /**
-     * Add a sample the sample store.
+     * Add a sample to the sample store, applying any filters.
      *
      * @param array $sample
      */
-    public function add_sample(\ExcimerLogEntry $sample): void {
-        $this->samples[] = $sample;
+    public function add_sample(\ExcimerLogEntry $sample) {
+        if (count($this->samples) === $this->samplelimit) {
+            $this->apply_doubling();
+        }
+        $this->counter += 1;
+        if ($this->counter === $this->filterrate) {
+            $this->samples[] = $sample;
+            $this->counter = 0;
+        }
+    }
```

Comment:
I'm not sure if it's exactly `1:1` with the `samplelimit_desc`, it does not have to be but it's a small detail that I'm not sure if it's worth fixing. It seems what is happening here is, the limit is reached, you are halving what you have, and you are adding a sample to the list. But this extra sample, would have been an entry from the _previous_ sample rate. To be more _correct_ do you think it should instead be returning early whenever the sample limit has been met?